mirascope/integrations/logfire/_utils.py

```diff
@@ -0,0 +1,180 @@
+"""Mirascope x Logfire Integration utils"""
+
+from contextlib import contextmanager
+from typing import (
+    Any,
+    Awaitable,
+    Callable,
+    Generator,
+    ParamSpec,
+    TypeVar,
+)
+
+import logfire
+from pydantic import BaseModel
+
+from ...core.base import BaseCallResponse
+from ...core.base._stream import BaseStream
+from ...core.base._structured_stream import BaseStructuredStream
+
+_P = ParamSpec("_P")
+_R = TypeVar("_R", bound=BaseCallResponse | BaseStream | BaseModel)
+
+
+@contextmanager
+def custom_context_manager(
+    fn: Callable[_P, _R] | Callable[_P, Awaitable[_R]],
+) -> Generator[logfire.LogfireSpan, Any, None]:
+    metadata = fn.__annotations__.get("metadata", {})
+    tags = getattr(metadata, "tags", {})
+    with logfire.with_settings(custom_scope_suffix="mirascope", tags=list(tags)).span(
+        fn.__name__
+    ) as logfire_span:
+        yield logfire_span
+
+
+def get_call_response_span_data(result: BaseCallResponse) -> dict:
+    return {
+        "async": False,
+        "call_params": result.call_params,
+        "model": result.model,
+        "provider": result._provider,
+        "prompt_template": result.prompt_template,
+        "template_variables": result.fn_args,
+        "messages": result.messages,
+        "response_data": result.response,
+    }
+
+
+def get_stream_span_data(stream: BaseStream) -> dict:
+    content = ""
+    if stream.message_param:
+        content = stream.message_param.get("content", "") or stream.message_param.get(
+            "message", ""
+        )
+    return {
+        "messages": [stream.user_message_param],
+        "call_params": stream.call_params,
+        "model": stream.model,
+        "provider": stream._provider,
+        "prompt_template": stream.prompt_template,
+        "template_variables": stream.fn_args,
+        "output": {"cost": stream.cost, "content": content},
```

Comment:
We should double check that things like `cost` aren't `None` to ensure we're actually replacing the iter properly (if you haven't already)

Reply:
i have confirmed they exist

Reply:
with stream_options include_usage set to true, both logfire and no logfire have streaming so iter works